Add documentation on OTEL support (#47194)
vercel/next.js: docs/advanced-features/instrumentation.md

---
description: Learn how to instrument your Next.js app.
---

> **Note**: This feature is experimental. To use it, you must explicitly opt in by defining `experimental.instrumentationHook = true;` in your `next.config.js`.

# `instrumentation.ts`

> If you export a function named `register` from this file, we will call that function whenever a new server instance is bootstrapped.

Reviewer: a new Next.js server

> When your `register` function is deployed, it will be called continuously on each cold boot (but exactly once in each environment).

Reviewer: this phrase is confusing

Author: Which exactly? line 9 or 10?

Author: Oh, you ment that `continuously` part. Removed that :+1:

Sometimes, it may be useful to import a file in your code because of the side effects it will cause. For example, you might import a file that defines a set of global variables, but never explicitly use the imported file in your code. You would still have access to the global variables the package has declared.

> You can import files with side effects in `instrumentation.ts`, which you might want to use in your `register` function as demonstrated in the following example:

Reviewer: add a mention that this file will get compiled for both node+ the edge, hence the usage of process.env for code pslitting

---

> description: Learn how instrument your Next.js app.

Reviewer 2: ```suggestion
description: Learn how to instrument your Next.js app.
```

---

> > **Note**: This feature is experimental, you need to explicitly opt-in by providing `experimental.instrumentationHook = true;` in your `next.config.js`.

Reviewer 2: ```suggestion
> **Note**: This feature is experimental. To use it, you must explicitly opt in by defining `experimental.instrumentationHook = true;` in your `next.config.js`.
```

# `instrumentation.ts`

> You can export `register` function from this file and we will be called whenever we bootstrap a new server instance.

Reviewer 2: ```suggestion
If you export a function named `register` from this file, Vercel will be call that function whenever we bootstrap a new server instance.
```

Feel free to ignore this suggestion if I'm diverting away from what you originally meant

> When deployed it will be called continuously on each coldboot (but exactly once in each environment).

Reviewer 2: ```suggestion
When your `register` function is deployed, it will be called continuously on each cold boot (but exactly once in each environment).
```

Is the `register` function what you're referring to when you say "it"? I totally understand if it seems redundant to keep being explicit here, but I do feel it helps when the subject is so technical. One of our docs goals is to make everything as approachable to juniors as possible

> Many people used imports with side-effects before we released this feature.

Reviewer 2: ```suggestion
Sometimes, it may be useful to import a file in your code because of the side effects it will cause. For example, you might import a file that defines a set of global variables, but never explicitly use the imported file in your code. You would still have access to the global variables the package has declared.

```

> The behavior is similar, but using imports with side-effects is hard to work with and has many down-sides.

Reviewer 2: ```suggestion
The behavior is similar, but using imports with side-effects is hard to work with and has many downsides.
```

What do you mean by "The behavior is similar" here? Are you referring to the behavior of `register`?

Author: Yes, I want to say that `register` does the same thing. But better and more robust/reliable.

> So instead of adding `import 'package-with-side-effect'`, you can add this:

Reviewer 2: ```suggestion
**We recommend importing the `init` function from `package-init` instead of importing packages with side effects**. You can then call it from your custom `register` function defined in `instrumentation.ts`, as demonstrated in the example below:
```

```ts
// instrumentation.ts

> import {init} from 'package-init'

Reviewer 2: ```suggestion
import { init } from 'package-init'
```

export const register() {
  init()
}
```

> If your package requires you to use import side-effect, you can do this instead:

Reviewer 2: ```suggestion
If you need to import a package strictly for its side effects, we recommend using `require` in your custom `register` function instead. The following example demonstrates a basic usage of `require` in a `register` function:
```

```ts
// instrumentation.ts

export const register() {
  require('package-with-side-effect')
}
```

> By doing this, you can have all your side-effects in one place and you can be sure that importing your files won't have unintended consequences.

Reviewer 2: ```suggestion
By doing this, you can colocate all of your side effects in one place in your code, and avoid any unintended consequences from importing files.
```

---
description: Learn how to instrument your Next.js app.
---

> **Note**: This feature is experimental. To use it, you must explicitly opt in by defining `experimental.instrumentationHook = true;` in your `next.config.js`.

# `instrumentation.ts`

> If you export a function named `register` from this file, we will call that function whenever new server instance is bootstrapped.

Reviewer 2: ```suggestion
If you export a function named `register` from this file, we will call that function whenever a new server instance is bootstrapped.
```

When your `register` function is deployed, it will be called continuously on each cold boot (but exactly once in each environment).

Sometimes, it may be useful to import a file in your code because of the side effects it will cause. For example, you might import a file that defines a set of global variables, but never explicitly use the imported file in your code. You would still have access to the global variables the package has declared.

The behavior is similar, but using imports with side-effects is hard to work with and has many downsides.

**We recommend importing the `init` function from `package-init` instead of importing packages with side effects**. You can then call it from your custom `register` function defined in `instrumentation.ts`, as demonstrated in the example below:

```ts
// instrumentation.ts

> import { init } from 'package-init'

Reviewer 2: What does `init` do?

Author: If the library needs initialization, it can be either through import side-effect or it exports a function that you can call. This should represent the second case.
